sklearn/model_selection/split.py

```diff
@@ -0,0 +1,1111 @@
+
+    Read more in the :ref:`User Guide <split>`.
+
+    Examples
+    --------
+    >>> from sklearn.model_selection import PredefinedSplit
+    >>> X = np.array([[1, 2], [3, 4], [1, 2], [3, 4]])
+    >>> y = np.array([0, 0, 1, 1])
+    >>> test_folds = [0, 1, -1, 1]
+    >>> ps = PredefinedSplit()
+    >>> ps.n_splits(X, y, labels=test_folds)
+    2
+    >>> print(ps)       # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    PredefinedSplit()
+    >>> for train_index, test_index in ps.split(X, y, labels=test_folds):
+    ...    print("TRAIN:", train_index, "TEST:", test_index)
+    ...    X_train, X_test = X[train_index], X[test_index]
+    ...    y_train, y_test = y[train_index], y[test_index]
+    TRAIN: [1 2 3] TEST: [0]
+    TRAIN: [0 2] TEST: [1 3]
+    """
+
+    def _iter_test_indices(self, X, y=None, labels=None):
+        test_fold, unique_folds = self._get_test_folds(X, y, labels)
+        for f in unique_folds:
+            yield np.where(test_fold == f)[0]
+
+    def _get_test_folds(self, X, y, labels):
+        test_fold = column_or_1d(np.array(labels, dtype=np.int))
+        unique_folds = np.unique(test_fold)
+        return test_fold, unique_folds[unique_folds != -1]
+
+    def n_splits(self, X, y=None, labels=None):
+        return len(self._get_test_folds(X, y, labels)[1])
+
+
+class CVIterator(_PartitionIterator):
+    """Wrapper class for old style cv objects and iterables."""
+    def __init__(self, cv):
+        self.cv = cv
+
+    def n_splits(self, X=None, y=None, labels=None):
+        """Returns the number of splitting iterations in the CV iterator"""
+        return len(self.cv)  # Both iterables and old-cv objects support len
+
+    def split(self, X=None, y=None, labels=None):
+        """Generate train/test indices to split data in train test sets.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training vectors, where n_samples is the number of samples
+            and n_features is the number of features.
+
+        y : array-like, shape (n_samples,)
+            The target variable for a supervised learning problem.
+
+        labels : array-like of int with shape (n_samples,), optional
+            Arbitrary domain-specific stratification of the data to be used
+            to draw the splits.
+        """
+        for train, test in self.cv:
+            yield train, test
+
+
+def check_cv(cv=3, y=None, classifier=False):
+    if cv is None:
+        cv = 3
+    if isinstance(cv, numbers.Integral):
+        if classifier and type_of_target(y) in ['binary', 'multiclass']:
+            return StratifiedKFold(cv)
+        else:
+            return KFold(cv)
+    if not hasattr(cv, 'split'):
+        if isinstance(cv, Iterable):
+            # Old style CV class / iterable
+            return CVIterator(cv)
+        else:
+            raise ValueError("The cv object, %s, is not an iterable to wrap."
```

Comment:
This branch will be reached even, for example, if passed a string or a float. (`cv=3.14`). The error message should therefore be friendlier. Maybe something like:

```
Expected cv as an integer, cross-validation generator (as in
sklearn.model_selection.split) or iterable, got: %s
```

This should be also clearly explained in the docstring. We should have a section in the narrative docs specifying the API for these cross-validation generators (the split signature) and link to that from here too.